### PR TITLE
feat(oracle): Parse multitable inserts

### PR DESCRIPTION
**Pull request by usefulalgorithm**
Adds support for parsing multitable insert statements in Oracle SQL. Test cases are converted from the official example SQL queries.

Ref: https://oracle-base.com/articles/9i/multitable-inserts

### CODE REVIEW

File: sqlglot/dialects/oracle.py [hunk 1]
```diff
@@ -226,6 +226,41 @@ def _parse_query_restrictions(self) -> t.Optional[exp.Expression]:
                 expression=self._match(TokenType.CONSTRAINT) and self._parse_field(),
             )
 
+        def _parse_multitable_inserts(self) -> exp.MultitableInserts:
+            kind = "first" if self._prev.token_type is TokenType.FIRST else "all"
+            expressions = []
+
+            def conditional_insert():
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:233]:
```suggestion
            def conditional_insert() -> t.Optional[exp.Conditionalinsert]:
```

File: sqlglot/dialects/oracle.py [hunk 2]
```diff
@@ -226,6 +226,41 @@ def _parse_query_restrictions(self) -> t.Optional[exp.Expression]:
...
+                expression = None
+                if self._match(TokenType.WHEN):
+                    expression = self._parse_assignment()
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:236]: Why is this `_parse_assignment`? I don't think we're expecting assignments in the condition, right?
```suggestion
                    expression = self._parse_disjunction()
```

File: sqlglot/dialects/oracle.py [hunk 3]
```diff
@@ -226,6 +226,41 @@ def _parse_query_restrictions(self) -> t.Optional[exp.Expression]:
                 expression=self._match(TokenType.CONSTRAINT) and self._parse_field(),
             )
 
+        def _parse_multitable_inserts(self) -> exp.MultitableInserts:
+            kind = "first" if self._prev.token_type is TokenType.FIRST else "all"
+            expressions = []
+
+            def conditional_insert():
+                expression = None
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:234]: Nit: let's set this to `None` using an `else` branch instead.

File: sqlglot/dialects/oracle.py [hunk 4]
```diff
@@ -226,6 +226,41 @@ def _parse_query_restrictions(self) -> t.Optional[exp.Expression]:
...
+                if self._match(TokenType.WHEN):
+                    expression = self._parse_assignment()
+                    self._match(TokenType.THEN)
+                else_matched = self._match(TokenType.ELSE)
+                else_ = else_matched is not None and else_matched
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:239]: Can't we just do `else_ = self._match(TokenType.ELSE)`?

File: sqlglot/dialects/oracle.py [hunk 5]
```diff
@@ -265,6 +300,7 @@ class Generator(generator.Generator):
             exp.Group: transforms.preprocess([transforms.unalias_group]),
             exp.ILike: no_ilike_sql,
             exp.Mod: rename_func("MOD"),
+            exp.MultitableInserts: lambda self, e: self.multitable_inserts_sql(e),
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:303]:
```suggestion
```

File: sqlglot/dialects/oracle.py [hunk 6]
```diff
@@ -328,3 +364,22 @@ def queryoption_sql(self, expression: exp.QueryOption) -> str:
         def coalesce_sql(self, expression: exp.Coalesce) -> str:
             func_name = "NVL" if expression.args.get("is_nvl") else "COALESCE"
             return rename_func(func_name)(self, expression)
+
+        def multitable_inserts_sql(self, expression: exp.MultitableInserts) -> str:
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:368]:
```suggestion
        def multitableinserts_sql(self, expression: exp.MultitableInserts) -> str:
```

File: sqlglot/dialects/oracle.py [hunk 7]
```diff
@@ -328,3 +364,22 @@ def queryoption_sql(self, expression: exp.QueryOption) -> str:
...
+            kind = self.sql(expression, "kind").upper()
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:369]: You won't need `upper` here if you apply the `self._prev.text.upper()` suggestion

File: sqlglot/dialects/oracle.py [hunk 8]
```diff
@@ -328,3 +364,22 @@ def queryoption_sql(self, expression: exp.QueryOption) -> str:
...
+            inserts = []
+            for conditional_insert in t.cast(t.List[exp.ConditionalInsert], expression.expressions):
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:371]: Why do we need a `t.cast` here?

File: sqlglot/dialects/oracle.py [hunk 9]
```diff
@@ -328,3 +364,22 @@ def queryoption_sql(self, expression: exp.QueryOption) -> str:
...
+                buffer = ""
+                if conditional_insert.expression:
+                    buffer += f"WHEN {self.sql(conditional_insert.expression)} THEN "
+                if conditional_insert.args["else_"]:
+                    buffer += "ELSE "
+                insert = t.cast(exp.Insert, conditional_insert.this)
+                insert_sql = self.sql(insert)
+                if insert_sql.startswith("INSERT "):
+                    insert_sql = insert_sql[len("INSERT ") :]
+                buffer += insert_sql
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:381]: This isn't right, you need to also add a method to generate `ConditionalInsert` and then use `self.expressions` or something.

File: sqlglot/dialects/oracle.py [hunk 10]
```diff
@@ -328,3 +364,22 @@ def queryoption_sql(self, expression: exp.QueryOption) -> str:
...
+                inserts.append(buffer.strip())
+
+            res = f"INSERT {kind} {' '.join(inserts)} {self.sql(expression.args['source'])}"
```
Comment (georgesittas) [inline sqlglot/dialects/oracle.py:384]:
```suggestion
            res = f"INSERT {kind} {' '.join(inserts)} {self.sql(expression, 'source')}"
```

**Review by georgesittas — COMMENTED** [review]
Thanks for the PR.

This will need more work to get merged, please look into my comments. Make sure to address the walrus operator issue and also run `make style` so this passes all tests.

File: sqlglot/parser.py [hunk 11]
```diff
@@ -2518,7 +2518,10 @@ def _parse_describe(self) -> exp.Describe:
             partition=partition,
         )
 
-    def _parse_insert(self) -> exp.Insert:
+    def _parse_multitable_inserts(self) -> exp.MultitableInserts:
+        raise NotImplementedError
```
[the remark this answers is not in the document]
Comment (usefulalgorithm) [reply sqlglot/parser.py:2522]: Only Oracle has multitable inserts, don't think we want to move it to the base parser?

[the remark this answers is not in the document]
Comment (georgesittas) [reply sqlglot/parser.py:2522]: Check out https://github.com/tobymao/sqlglot/pull/4008#discussion_r1736400105

File: sqlglot/dialects/oracle.py (hunk 8 above)
Comment (usefulalgorithm) [reply sqlglot/dialects/oracle.py:371]: We don't, `t.cast` just makes IDE show `expression.expressions` as a `List[exp.ConditionalInsert]`. I don't think it has any real effect.

Comment (georgesittas) [reply sqlglot/dialects/oracle.py:371]: Let's remove it then please.

File: sqlglot/parser.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (usefulalgorithm) [reply sqlglot/parser.py:2522]: SG, thanks

**Pull request merged**